app/src/main/java/ca/vijayan/flyweb/proxy/ProxyServer.java

```diff
@@ -216,8 +211,23 @@ void handleLocalRead(ProxyConnection conn) {
             handleLocalIncomingData(conn, totalRead);
         }
     }
-    void handleLocalIncomingData(ProxyConnection conn, int nbytes) {
-        // TODO: IMPLEMENT.
+
+    void handleLocalIncomingData(final ProxyConnection conn, final int nbytes) {
+        ByteBuffer buf = conn.getLocalRecvBuffer();
+        byte[] data = new byte[buf.remaining()];
+        buf.get(data);
+
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_FIXED_THREAD_POOL);
+        final byte[] DATA = data;
+
+        Future<?> future = executor.submit(new Callable<Void>() {
+            public Void call() {
+                mProxyDataHandler.handleLocalDataReceived(conn, DATA);
+                return null;
+            }
+        });
+
+        getFromFuture(future, "Failed to handle local incoming data.");
```

Comment:
I don't think we should be blocking here.  In fact, I don't think the IO thread needs to care at all about waiting on the callable's execution.  Anything that needs to be done can be shoved off into handleLocalDataReceived() implementations.